app/services/fact_importer.rb

```diff
@@ -37,9 +37,11 @@ def initialize(host, facts = {})
 
   # expect a facts hash
   def import!
-    delete_removed_facts
-    add_new_facts
-    update_facts
+    ActiveRecord::Base.transaction do
+      delete_removed_facts
+      add_new_facts
+      update_facts
```

Comment:
I think you should switch the order to first update then add, otherwise you are going to check if the new facts were changed (which they obviously have not) during the update.